modules/connectivity/TrackStreamingStatus.ts

```diff
@@ -0,0 +1,683 @@
+import browser from '../browser';
+import Statistics from '../statistics/statistics';
+
+/** Track streaming statuses. */
+type TrackStreamingStatus = 'active' | 'inactive' | 'interrupted' | 'restoring';
+interface TrackStreamingStatusMapType {
+    [key: string]: TrackStreamingStatus
+}
+
+type VideoType = 'camera' | 'desktop';
+type StreamingStatusMap = {
+    videoType?: VideoType,
+    startedMs?: number,
+    p2p?: boolean,
+    streamingStatus?: string,
+    value?: number
+};
+
+const logger = getLogger(__filename);
+
+/**
+ * Default value of 500 milliseconds for {@link TrackStreamingStatusImpl.outOfForwardedSourcesTimeout}.
+ */
+const DEFAULT_NOT_IN_FORWARDED_SOURCES_TIMEOUT = 500;
+
+/**
+ * Default value of 2500 milliseconds for {@link TrackStreamingStatusImpl.p2pRtcMuteTimeout}.
+ */
+const DEFAULT_P2P_RTC_MUTE_TIMEOUT = 2500;
+
+/**
+ * Default value of 10000 milliseconds for {@link TrackStreamingStatusImpl.rtcMuteTimeout}.
+ */
+const DEFAULT_RTC_MUTE_TIMEOUT = 10000;
+
+/**
+ * The time to wait a track to be restored. Track which was out of forwarded sources should be inactive and when
+ * entering forwarded sources it becomes restoring and when data is received from bridge it will become active, but if
+ * no data is received for some time we set status of that track streaming to interrupted.
+ */
+const DEFAULT_RESTORING_TIMEOUT = 10000;
+
+export const TrackStreamingStatusMap: TrackStreamingStatusMapType = {
+    /**
+     * Status indicating that streaming is currently active.
+     */
+    ACTIVE: 'active',
+
+    /**
+     * Status indicating that streaming is currently inactive.
+     * Inactive means the streaming was stopped on purpose from the bridge, like exiting forwarded sources or
+     * adaptivity decided to drop video because of not enough bandwidth.
+     */
+    INACTIVE: 'inactive',
+
+    /**
+     * Status indicating that streaming is currently interrupted.
+     */
+    INTERRUPTED: 'interrupted',
+
+    /**
+     * Status indicating that streaming is currently restoring.
+     */
+    RESTORING: 'restoring'
+};
+
+/**
+ * Class is responsible for emitting JitsiTrackEvents.TRACK_STREAMING_STATUS_CHANGED events.
+ */
+export class TrackStreamingStatusImpl {
+    rtc: RTC;
+    conference: JitsiConference;
+    track: JitsiRemoteTrack;
+
+    /**  This holds the timeout callback ID scheduled using window.setTimeout. */
+    trackTimer: number | null;
+
+    /**
+     * If video track frozen detection through RTC mute event is supported, we wait some time until video track is
+     * considered frozen. But because when the track falls out of forwarded sources it is expected for the video to
+     * freeze this timeout must be significantly reduced in "out of forwarded sources" case.
+     *
+     * Basically this value is used instead of {@link rtcMuteTimeout} when
+     * track is not in forwarded sources.
+     */
+    outOfForwardedSourcesTimeout: number;
```

Comment:
Something I've noticed is that after changing this file's extension to .ts and running `npm run build`, all the class properties types were removed in the auto generated type declarations file. I had to explicitly define them here